suggestions from IBM Research collaborators

walkup opened the pull request:
There are a number of suggested code changes based on our experience with the SBD code on x86 servers.  The main ones are briefly described below.

(1) In the mult() routine, mult.h, OpenMP schedule(runtime) => schedule(dynamic).  We really want "dynamic" thread scheduling for OpenMP, so it is best to specify that in the code.  With the "runtime" schedule, one can get "dynamic" thread scheduling, but only if the user specifies the correct OpenMP environment variable in the job script.

(2) There is a new implementation of DetFromAlpbaBeta() in determinants.h.  This one is significantly faster ... please check it out.

(3) MPI-4 routines, MPI_Isend_c, MPI_Irecv_c, and MPI_Allreduce_c were added to mpi_utility.h and out_of_place/davidson.h with a pre-processor test to check for MPI-4 support.  That way the MPI-4 routines will get used if the MPI implementation supports them, while the older MPI routines MPI_Isend, MPI_Irecv, MPI_Allreduce will be used if the MPI version is less than 4.  This code should work with any MPI version.

(4) When we tried 10^11 subspace dimension, we noticed significant memory allocation in the GenerateExcitations() function in tpb/helper.h.  We modified the code to first count the number of single and double excitations, and then allocate only the memory that is really needed for the excitation lists.  That approach saves a lot of memory and results in more compact lists.

(5) For very large subspace dimensions it can take a long time to get reduced density-matrix output via the --rdm command-line option.  We added OpenMP to one loop and changed the OpenMP work assignment from block-partition to round-robin in the Correlation() routine, tpb/correlation.h to speed that up.

Those are the main suggestions.

walkup committed:
new DetFromAlphaBeta() in determinants.h, OpenMP schedule(dynamic) in mult.h, save memory in GenerateExcitation() helper.h

walkup committed:
removed unused variables in determinants.h and helper.h

walkup committed:
use MPI-4 routine MPI_Allreduce_c() instead of MPI_Allreduce() in basic/out_of_place_func/davidson.h and framework/mpi_utility.h

walkup committed:
use MPI-4 routines MPI_Isend_c() MPI_Irecv_c() in Mpi2dSlide to avoid integer overflow

walkup committed:
restore alternate OpenMP version of the mult() routine using round-robin work assignment in mult.h

walkup committed:
check MPI version for MPI-4 support in mpi_utility.h

walkup committed:
change work assignment to OpenMP threads from block to round-robin in correlation.h

walkup committed:
updated Makefile and Configuration

walkup committed:
added OpenMP parallel region to Correlation() in correlation.h

walkup committed:
added MPI version check in basic/out_of_place_func/davidson.h for MPI-4 support

walkup committed:
added fcidump and alpha bit-strings for N2 in the data/n2 directory

doichanj commented:
It seems the optimized `DetFromAlpbaBeta` in this PR is applied for `SBD_TRADMODE` (I think these sections are for trad mode of Fujitsu's compiler for Fugaku), is it right?

walkup commented:
The code has the pre-processor check "#ifdef SBD_TRADMODE" in two places : determinants.h and mult.h.  In practice one needs to compile with -DSBD_TRADMODE in CCFLAGS in order to select the optimized code path ... so that is independent of the choice of compiler or other options.  Without the -DSBD_TRADMODE option, one gets the original code for DetFromAlphaBeta() in determinants.h, and one gets a round-robin assignment of work to threads for the mult() routine in mult.h.  Note that the round-robin thread assignment is already a substantial improvement over the original code, which used a static block partition of work, which often resulted in more load imbalance and correspondingly lower speed-up with increasing number of OpenMP threads.  The suggested change to the optimized mult() routine is only to specify schedule(dynamic) instead of schedule(runtime).  That way an end-user will get dynamic scheduling without the need to set env variable OMP_SCHEDULE=dynamic in the job script.  It happens that the GNU OpenMP runtime uses dynamic scheduling by default when one specifies schedule(runtime), but that behavior is not guaranteed by standards, and other OpenMP implementations may use the static block partitioning method by default.  The safe way to ensure dynamic scheduling is to specify it in the code.

I have a suggestion for performance tests : use the fcidump and alphadets files for the 3.6e7 subspace dimension problem ... 6000 alpha bit-strings.  In order to focus on compute performance, it would suffice to use one node, one MPI rank and 48 OpenMP threads.  One "iteration" is sufficient for a short performance test.  This would take MPI out of the picture.  Of course MPI scaling is critical in practice, but that is a different issue.  When the code is compiled with -DSBD_TRADMODE, it will pick up an optimized DetFromAlphaBeta() routine using pre-computed index lists.  That already speeds up DetFromAlphaBeta() quite a bit.  The new implementation of DetFromAlphaBeta() in the pull request is a little faster in our tests, but the performance difference will not be very much when comparing with the code already included with -DSBD_TRADMODE.  It is best to compare with and without -DSBD_TRADMODE.

The pull request has quite a few suggestions for code updates that are not performance related. Examples are listed here : 

(1) Processing the --max_time command-line argument was added to tpb/sbdiag.h.  With the original code, a --max_time argument set in the job script would just be ignored.

(2) When attempting to load a wavefunction file as input, the original code would simply continue and generate NANs if the input file was not found ... so I added error messages and MPI_Aborts in  tpb/restart.h.

(3) For very large problems, memory allocation for the lists of one-electron and two-electron excitations can be an issue.  I added code to allocate only the amount of memory that is required, instead of allocating the maximum amount for both lists ... in GenerateExcitation(), helper.h.

There are suggested code changes that affect OpenMP threading in the Correlation() routine, which is only called if one requests reduced density-matrix output via command-line option --rdm 1.  The time spent in that section of code is reported as "time for measurement".  Our experience has been that this code section can take quite a long time when the subspace dimension is very large.

It might be convenient to use my fork of this repository for testing purposes : https://github.com/walkup/sbd-ykt.git  .

doichanj commented:
I think it will be better to keep `SBD_TRADMODE` as is and make other option to enable optimization for this PR or insert in the non-`SBD_TRADMODE` sections

doichanj commented:
I tested this PR on Fugaku with and without `SBD_TRADMODE`
(But compiled with clang mode for both cases)

here is script to run
```
#!/bin/sh
#PJM -L "node=96"
#PJM -L "rscgrp=q-IBM-M"
#PJM -L "elapse=3:00:00"
#PJM -g hp240496
#PJM --mpi "max-proc-per-node=4"
#PJM -x PJM_LLIO_GFSCACHE=/vol0002:/vol0004
#PJM --llio sharedtmp-size=20Gi
#PJM --llio localtmp-size=60Gi
#PJM -s

export OMP_NUM_THREADS=12

mpiexec -n 384 /vol0206/data/ra010014/u13452/qcsc/sbd-ykt/samples/selected_basis_diagonalization/diag --fcidump /vol0206/data/ra010014/u13452/qcsc/sbd-ykt/samples/selected_basis_diagonalization/fcidump.txt --adetfile /vol0206/data/ra010014/u13452/qcsc/sbd-ykt/samples/selected_basis_diagonalization/AlphaDets.bin --method 0 --block 10 --iteration 30 --tolerance 1.0e-4 --eps 1.0e-4 --adet_comm_size 4 --bdet_comm_size 4 --task_comm_size 4 --init 0 --shuffle 0 --carryover_ratio 0.5 --savename wf --carryoverfile carryover.txt --rdm 0 --max_time 24000
```

here is an output without `SBD_TRADMODE`
```
 Elapsed time for helper construction 0.268677 (sec) 
 Elapsed time for init 0.001547 (sec) 
 Davidson iteration 0.0 (tol=0.5584049335932896): -326.2408941280003
 Davidson iteration 0.1 (tol=0.5273981009133736): -326.2727415524768 -326.2091473255696
 Davidson iteration 0.2 (tol=0.388204457239284): -326.4178430437822 -326.2181988615177 -326.0172122206592
 Davidson iteration 0.3 (tol=0.3497729713517192): -326.4713519238916 -326.3557367486418 -326.2181319105965 -326.0098087398594
 Davidson iteration 0.4 (tol=0.2808618773151046): -326.5832547466651 -326.3728581604678 -326.2488834558872 -326.2180261078601
 Davidson iteration 0.5 (tol=0.198733256359907): -326.6733247990203 -326.3827196556921 -326.254214642412 -326.2181261048361
 Davidson iteration 0.6 (tol=0.1424807026850615): -326.717065103882 -326.4076843308267 -326.2744276287287 -326.218708469311
 Davidson iteration 0.7 (tol=0.09859219297260297): -326.7372370812941 -326.4505003459511 -326.3130582992134 -326.2191884205282
 Davidson iteration 0.8 (tol=0.07931361945032463): -326.7479252638605 -326.5287199295561 -326.3509495832776 -326.2195689809966
 Davidson iteration 0.9 (tol=0.07419503855500596): -326.75528163037 -326.6045188660332 -326.3840757736334 -326.2207429243194
 Davidson iteration 1.0 (tol=0.07419503855500859): -326.7552816303472
 Davidson iteration 1.1 (tol=0.05140028195415903): -326.7588550928672 -325.6476554008519
 Davidson iteration 1.2 (tol=0.0670041635116501): -326.7633688330964 -326.5263805941568 -324.6727448376425
 Davidson iteration 1.3 (tol=0.08017598157450892): -326.7699572582404 -326.6839862147834 -325.3343254154428 -324.5798322956593
 Davidson iteration 1.4 (tol=0.07117575513543181): -326.7778571648172 -326.7243090870052 -325.8671417119532 -324.8959453820922
 Davidson iteration 1.5 (tol=0.05337071676127358): -326.782742615209 -326.7355446874851 -326.1993208165532 -325.3611160235101
 Davidson iteration 1.6 (tol=0.04595740421107942): -326.7858027869378 -326.7410707518957 -326.4419082138867 -325.7038866073128
 Davidson iteration 1.7 (tol=0.05131953226046268): -326.7888658822574 -326.746403505193 -326.6066017205891 -326.0200075655171
 Davidson iteration 1.8 (tol=0.06757810357598019): -326.7938282009554 -326.7555173758349 -326.6943962429306 -326.2344352680985
 Davidson iteration 1.9 (tol=0.05841650860721943): -326.7983220873277 -326.7631261326055 -326.7163566232279 -326.3337459669934
 Davidson iteration 2.0 (tol=0.05841650860727956): -326.7983220873355
 Davidson iteration 2.1 (tol=0.04343399920585904): -326.8005272628686 -325.6989202154597
 Davidson iteration 2.2 (tol=0.04548367796850045): -326.8032473056741 -326.5305444922893 -324.640427093645
 Davidson iteration 2.3 (tol=0.05122468478780325): -326.8061931393706 -326.6930326491574 -325.6960109727837 -324.3988481026421
 Davidson iteration 2.4 (tol=0.05545919784909142): -326.8097182236088 -326.7492576997347 -326.0262730170614 -325.2668972080129
 Davidson iteration 2.5 (tol=0.05462369672609727): -326.8134696315619 -326.7716327931915 -326.2773427681497 -325.6737432736398
 Davidson iteration 2.6 (tol=0.04528226600676757): -326.8164141825677 -326.7807957212207 -326.4447332692444 -325.8162148582715
 Davidson iteration 2.7 (tol=0.04121761600391612): -326.8186074125443 -326.7856988480205 -326.5643711588443 -326.0240825954571
 Davidson iteration 2.8 (tol=0.03267148913820877): -326.8202079888685 -326.7887992206051 -326.6406668136262 -326.1975608276073
 Davidson iteration 2.9 (tol=0.03376459665189684): -326.8212990277959 -326.7909271694315 -326.68825721939 -326.334715525066
 Davidson iteration 3.0 (tol=0.03376459665191736): -326.8212990277918
 Davidson iteration 3.1 (tol=0.02138891533081055): -326.8219515957762 -325.5264081466293
 Davidson iteration 3.2 (tol=0.02443380290512191): -326.8226057277935 -326.522138514918 -324.5506516989606
 Davidson iteration 3.3 (tol=0.02438562750284775): -326.8232937352757 -326.7061578789307 -325.1908004234983 -324.4378067125765
 Davidson iteration 3.4 (tol=0.02438881517101366): -326.8240048975371 -326.7628725044524 -325.8261803656382 -324.7892325274544
 Davidson iteration 3.5 (tol=0.02119646535646047): -326.8246376890716 -326.7841438355145 -326.2030074578869 -325.1827134367816
 Davidson iteration 3.6 (tol=0.01728134503985551): -326.8250842727468 -326.793191805236 -326.4088940520842 -325.5626201196674
 Davidson iteration 3.7 (tol=0.01410112168643186): -326.8253991665229 -326.7979418416892 -326.5426613379253 -325.9693329233559
 Davidson iteration 3.8 (tol=0.01108794359857416): -326.8256092585518 -326.8007463846032 -326.62805570437 -326.2069175940953
 Davidson iteration 3.9 (tol=0.01045088626713294): -326.8257148729838 -326.8021284371927 -326.6713446622979 -326.3308399553111
 Davidson iteration 4.0 (tol=0.01045088626715063): -326.825714872988
 Davidson iteration 4.1 (tol=0.007894392447088807): -326.8257857500774 -325.6993830915507
 Davidson iteration 4.2 (tol=0.007818293586323672): -326.8258689303879 -326.5388388881354 -324.5587491350623
 Davidson iteration 4.3 (tol=0.007252443000638754): -326.8259387136601 -326.693009463162 -325.6039145888167 -324.2970495020753
 Davidson iteration 4.4 (tol=0.006529475887386034): -326.8259954031722 -326.746820502964 -325.9677562844869 -325.1756535163057
 Davidson iteration 4.5 (tol=0.005658501932827411): -326.826039616292 -326.7708892074642 -326.2458927831289 -325.5695016728428
 Davidson iteration 4.6 (tol=0.004576097799135509): -326.8260699350485 -326.7823625383714 -326.4263298408509 -325.769089783662
 Davidson iteration 4.7 (tol=0.004040609801457541): -326.8260911278986 -326.7892660264429 -326.5524871809623 -326.0083001634078
 Davidson iteration 4.8 (tol=0.003266057081586567): -326.8261067159673 -326.7939472959982 -326.6332196762434 -326.1844414150623
 Davidson iteration 4.9 (tol=0.004096062329645446): -326.8261191644589 -326.7978139424191 -326.6894862032323 -326.3384760209339
 Davidson iteration 5.0 (tol=0.004096062329651331): -326.8261191644582
 Davidson iteration 5.1 (tol=0.002407013502181484): -326.8261280282292 -325.4083613867979
 Davidson iteration 5.2 (tol=0.00254498387856309): -326.8261355089276 -326.4646348897735 -324.5427765127401
 Davidson iteration 5.3 (tol=0.002432631222165158): -326.8261425001571 -326.6819027319738 -325.1180804801385 -324.440729064627
 Davidson iteration 5.4 (tol=0.002425569683854807): -326.8261493415111 -326.7521648868918 -325.7791675412254 -324.7482045415022
 Davidson iteration 5.5 (tol=0.002281262878099328): -326.8261559046801 -326.7806388777461 -326.189693122031 -325.1002011312816
 Davidson iteration 5.6 (tol=0.002115527784771208): -326.826161616537 -326.7940841935528 -326.4157419756718 -325.5105172929906
 Davidson iteration 5.7 (tol=0.001991317073709009): -326.826166781498 -326.8017661301923 -326.553412997124 -325.9238367165777
 Davidson iteration 5.8 (tol=0.00187238536940742): -326.8261715823226 -326.8068376940398 -326.6409563857352 -326.1880544643249
 Davidson iteration 5.9 (tol=0.002289208022783787): -326.8261753111692 -326.8099177170562 -326.6889412571651 -326.3351250781727
 Davidson iteration 6.0 (tol=0.002289208022725868): -326.8261753111664
 Davidson iteration 6.1 (tol=0.001584230506308963): -326.8261783814856 -325.6040818201074
 Davidson iteration 6.2 (tol=0.001603571841705376): -326.8261816924779 -326.5202459484846 -324.5221457788356
 Davidson iteration 6.3 (tol=0.001691713113326594): -326.8261849582121 -326.7037647836821 -325.5616376453633 -324.2945705674285
 Davidson iteration 6.4 (tol=0.001896336158437876): -326.8261887366872 -326.7671944057417 -325.9621126658236 -325.1424579719683
 Davidson iteration 6.5 (tol=0.002266190682748636): -326.8261938370149 -326.7954637469631 -326.2754050514819 -325.5159917695902
 Davidson iteration 6.6 (tol=0.002800275281502448): -326.8262011720957 -326.8092384972177 -326.4683797216223 -325.7361143456834
 Davidson iteration 6.7 (tol=0.00399027365347491): -326.8262139200324 -326.8171499830616 -326.5893889603415 -325.9924703508534
 Davidson iteration 6.8 (tol=0.006333293070394772): -326.8262437739975 -326.8220344894751 -326.6617266558742 -326.1867744561258
 Davidson iteration 6.9 (tol=0.02651030646236624): -326.8264079636891 -326.8251772284255 -326.7095414498882 -326.3549001889718
 Davidson iteration 7.0 (tol=0.02651030646236862): -326.8264079636918
 Davidson iteration 7.1 (tol=0.01446067324456974): -326.8267563160144 -325.327429534191
 Davidson iteration 7.2 (tol=0.01377810885837969): -326.8269973817596 -326.40160083392 -324.5465643244038
 Davidson iteration 7.3 (tol=0.01251143881894542): -326.8271925062136 -326.6543032720368 -325.0829271570734 -324.4551361789343
 Davidson iteration 7.4 (tol=0.01187193743617107): -326.8273632629166 -326.7371997116032 -325.7320576599345 -324.737637624852
 Davidson iteration 7.5 (tol=0.01119734970011098): -326.8275206286926 -326.772812088118 -326.1711402694953 -325.0693865290228
 Davidson iteration 7.6 (tol=0.01073603046148396): -326.8276620344819 -326.7904846464231 -326.413301298108 -325.4683435290239
 Davidson iteration 7.7 (tol=0.01059707262251217): -326.8277986944636 -326.8010617876563 -326.554468138736 -325.8759405182385
 Davidson iteration 7.8 (tol=0.01035118653364241): -326.8279408521935 -326.8082345886612 -326.6411521484723 -326.1557526962051
 Davidson iteration 7.9 (tol=0.01424410315562168): -326.8280674994808 -326.8126108252224 -326.6906090767121 -326.3258560144757
 Davidson iteration 8.0 (tol=0.01424410315562347): -326.8280674994814
 Davidson iteration 8.1 (tol=0.00991276002939729): -326.8281867370978 -325.6278787495811
 Davidson iteration 8.2 (tol=0.01062081243075205): -326.8283251541746 -326.5562421878427 -324.4955503643819
 Davidson iteration 8.3 (tol=0.01120940127905132): -326.8284669966865 -326.7209937024257 -325.5208979039675 -324.2735384259726
 Davidson iteration 8.4 (tol=0.01205763137443869): -326.8286254844582 -326.7750708143419 -325.9437434499411 -325.1078141340993
 Davidson iteration 8.5 (tol=0.013095891576845): -326.8288121739156 -326.7983003042265 -326.272516957372 -325.4624100529032
 Davidson iteration 8.6 (tol=0.01386737126028106): -326.829021287653 -326.8094243507331 -326.468707642951 -325.6854410451898
 Davidson iteration 8.7 (tol=0.01491542446988013): -326.8292562956503 -326.8155879618495 -326.5881834937337 -325.9536593745557
 Davidson iteration 8.8 (tol=0.01456056715222466): -326.8295167703953 -326.8193076116092 -326.6611381001463 -326.1670045090902
 Davidson iteration 8.9 (tol=0.02218950942314267): -326.8298158897269 -326.8218087213784 -326.7117040126858 -326.354032620387
 Davidson iteration 9.0 (tol=0.02218950942314603): -326.8298158897228
 Davidson iteration 9.1 (tol=0.01233312163245378): -326.8300642113787 -325.3679976886928
 Davidson iteration 9.2 (tol=0.01204411935990142): -326.8302446214724 -326.426908290613 -324.5477069264207
 Davidson iteration 9.3 (tol=0.01086098868439595): -326.8303912579059 -326.6645387139398 -325.061683621425 -324.4570171439842
 Davidson iteration 9.4 (tol=0.009827007168369954): -326.8305125052748 -326.7400407480273 -325.7046356399994 -324.728892225862
 Davidson iteration 9.5 (tol=0.008632993269792583): -326.8306117850295 -326.7718919269948 -326.1503738758767 -325.0419015920268
 Davidson iteration 9.6 (tol=0.007557333110888706): -326.8306874950127 -326.7874505844544 -326.3971143828325 -325.4298055131591
 Davidson iteration 9.7 (tol=0.006638170463940518): -326.8307464028443 -326.7966335455417 -326.5389830334167 -325.8278275491376
 Davidson iteration 9.8 (tol=0.005460990566546598): -326.8307935109619 -326.8026925900743 -326.6254330778983 -326.1159100676932
 Davidson iteration 9.9 (tol=0.006862474709272067): -326.8308251630896 -326.8064128970836 -326.6783209945635 -326.3079693992182
 Davidson iteration 10.0 (tol=0.006862474709224072): -326.8308251630857
 Davidson iteration 10.1 (tol=0.004968448980168143): -326.8308536169053 -325.6525053178546
 Davidson iteration 10.2 (tol=0.005727890727516652): -326.8308910842559 -326.584135235384 -324.4729683203631
 Davidson iteration 10.3 (tol=0.006071252112425417): -326.8309317328756 -326.7330617251665 -325.4849944160765 -324.2543184592753
 Davidson iteration 10.4 (tol=0.006322401678805121): -326.8309761699928 -326.780257187207 -325.9245482292429 -325.0777007393423
 Davidson iteration 10.5 (tol=0.006418348849862786): -326.8310235254464 -326.8000627295974 -326.2656162151779 -325.4178644119741
 Davidson iteration 10.6 (tol=0.006293143118906995): -326.8310694260107 -326.8096062650064 -326.4667385313275 -325.6460382050923
 Davidson iteration 10.7 (tol=0.006159568751851669): -326.8311129245533 -326.8149591462747 -326.5870803160669 -325.9224194216236
 Davidson iteration 10.8 (tol=0.005499205957664407): -326.8311533792615 -326.818290282307 -326.6606646941643 -326.1486102936726
 Davidson iteration 10.9 (tol=0.007938205266401193): -326.8311933093007 -326.8206789313792 -326.7125000308379 -326.3488142195183
 Davidson iteration 11.0 (tol=0.007938205266410137): -326.831193309312
 Davidson iteration 11.1 (tol=0.004426110267180962): -326.8312250629221 -325.3686793781839
 Davidson iteration 11.2 (tol=0.004348867633352149): -326.8312483298745 -326.4272380714915 -324.5496608215708
 Davidson iteration 11.3 (tol=0.003926281232965237): -326.8312672790134 -326.6647017210259 -325.0424774809977 -324.4601071643591
 Davidson iteration 11.4 (tol=0.003523630201997803): -326.8312828090483 -326.7394714134215 -325.6846671709913 -324.7209732397241
 Davidson iteration 11.5 (tol=0.003080535679690207): -326.8312953586895 -326.7710417251116 -326.1390245477113 -325.0175229126284
 Davidson iteration 11.6 (tol=0.002673877617127081): -326.8313047971409 -326.7864431795721 -326.3907108720674 -325.4030924899547
 Davidson iteration 11.7 (tol=0.002308892714029646): -326.8313119199198 -326.7954434339139 -326.5332822402879 -325.7966888040519
 Davidson iteration 11.8 (tol=0.001839800825237995): -326.8313174006445 -326.8012385986557 -326.6195179037617 -326.0919836630983
 Davidson iteration 11.9 (tol=0.002314769080805312): -326.8313209505 -326.8047513805308 -326.6739567124046 -326.2982286020277
 Davidson iteration 12.0 (tol=0.002314769080831606): -326.831320950488
 Davidson iteration 12.1 (tol=0.001683968047855768): -326.83132416869 -325.6389457598092
 Davidson iteration 12.2 (tol=0.001986496381937429): -326.8313285362947 -326.5860228487903 -324.459489124442
 Davidson iteration 12.3 (tol=0.002118479461276679): -326.831333385249 -326.7344141836778 -325.4552227489585 -324.2472727351758
 Davidson iteration 12.4 (tol=0.002201943442659391): -326.8313387284319 -326.7809330167801 -325.9062799420901 -325.0539462111964
 Davidson iteration 12.5 (tol=0.002219280092175037): -326.8313443734103 -326.800281709183 -326.2586043027408 -325.3828272059568
 Davidson iteration 12.6 (tol=0.002163385154256533): -326.8313497761449 -326.8096146805414 -326.4645413793997 -325.613596234041
 Davidson iteration 12.7 (tol=0.002090606651635489): -326.8313548073706 -326.8148395468145 -326.5861486065476 -325.8966400823351
 Davidson iteration 12.8 (tol=0.001848672529147346): -326.8313594099801 -326.8180993826717 -326.6605556623723 -326.1341041233269
 Davidson iteration 12.9 (tol=0.002684683405775689): -326.8313639243252 -326.8204676565496 -326.7133840213648 -326.3446125578338
 Davidson iteration 13.0 (tol=0.002684683405760038): -326.8313639243281
 Davidson iteration 13.1 (tol=0.00149652111389192): -326.8313675449895 -325.3666060339796
 Davidson iteration 13.2 (tol=0.00147627200492255): -326.8313702029766 -326.4263391345108 -324.5529931296331
 Davidson iteration 13.3 (tol=0.001338450178668994): -326.8313723799755 -326.6649078285193 -325.0272334934452 -324.4641288224078
 Davidson iteration 13.4 (tol=0.001200356305964475): -326.8313741688818 -326.7395396880834 -325.6698474935736 -324.7161256931164
 Davidson iteration 13.5 (tol=0.001050972067100083): -326.8313756152929 -326.7709716865609 -326.1308201346263 -324.9980402821917
 Davidson iteration 13.6 (tol=0.0009127081146999037): -326.8313767043014 -326.7862970377458 -326.3869689816172 -325.382819372718
 Davidson iteration 13.7 (tol=0.0007843678621576052): -326.8313775198911 -326.7951966497911 -326.5305506911856 -325.7735210583299
 Davidson iteration 13.8 (tol=0.0006177990240089032): -326.831378141369 -326.8008671637189 -326.6170121927141 -326.0748615620674
 Davidson iteration 13.9 (tol=0.0007837555069094398): -326.8313785416305 -326.804287979215 -326.6725537527144 -326.2916892672791
 Davidson iteration 14.0 (tol=0.0007837555068564781): -326.8313785416344
 Davidson iteration 14.1 (tol=0.0005697460966853757): -326.8313789077527 -325.629105094266
 Davidson iteration 14.2 (tol=0.0006793767266081498): -326.8313794095401 -326.5860681771376 -324.4491605559806
 Davidson iteration 14.3 (tol=0.0007275586755016729): -326.8313799731214 -326.7349697399018 -325.4300655723303 -324.2434812163768
 Davidson iteration 14.4 (tol=0.0007563718814021683): -326.8313805984304 -326.7812973904416 -325.8905182543092 -325.0335527850962
 Davidson iteration 14.5 (tol=0.0007614449535919533): -326.8313812587947 -326.800458683365 -326.2524586968424 -325.3538422145022
 Davidson iteration 14.6 (tol=0.0007420411466831559): -326.8313818900363 -326.8096928796618 -326.462444290145 -325.5867480464556
 Davidson iteration 14.7 (tol=0.0007132871129740408): -326.8313824749818 -326.8148471189088 -326.5851959128096 -325.8753710773218
 Davidson iteration 14.8 (tol=0.0006281948604979734): -326.8313830075086 -326.8180593530279 -326.660310647401 -326.1223138546727
 Davidson iteration 14.9 (tol=0.0009185741620302342): -326.8313835303629 -326.8204069797044 -326.7139016960695 -326.3410012062939
 Davidson iteration 15.0 (tol=0.0009185741620342883): -326.831383530363
 Davidson iteration 15.1 (tol=0.0005120057687082147): -326.8313839532694 -325.3661281683563
 Davidson iteration 15.2 (tol=0.0005067445192115045): -326.8313842642899 -326.4261477388512 -324.5563973160699
 Davidson iteration 15.3 (tol=0.0004608898834595038): -326.8313845201702 -326.6653661066206 -325.014657572067 -324.467987808616
 Davidson iteration 15.4 (tol=0.0004131029334642481): -326.8313847308916 -326.7398177190576 -325.6579078440446 -324.7128394169526
 Davidson iteration 15.5 (tol=0.0003619878855852615): -326.8313849012587 -326.771091875787 -326.1240804915914 -324.9819833732147
 Davidson iteration 15.6 (tol=0.0003146304554459176): -326.8313850296145 -326.7863398807478 -326.3840560428997 -325.3663743335218
 Davidson iteration 15.7 (tol=0.0002696657148730309): -326.8313851253383 -326.795165051588 -326.5287369913747 -325.7550485539862
 Davidson iteration 15.8 (tol=0.0002108581368080578): -326.8313851979184 -326.8007606987852 -326.6155722422914 -326.0613114489375
 Davidson iteration 15.9 (tol=0.0002693031347559761): -326.8313852445756 -326.8041313904845 -326.6719406848798 -326.2864357238008
 Davidson iteration 16.0 (tol=0.0002693031347272566): -326.8313852445747
 Davidson iteration 16.1 (tol=0.0001955007504026613): -326.8313852875202 -325.6222797951286
 Davidson iteration 16.2 (tol=0.0002346116767919264): -326.8313853466397 -326.5860133414091 -324.44039014484
 Davidson iteration 16.3 (tol=0.000251800895887373): -326.8313854134344 -326.7353572988375 -325.4081998408335 -324.2406741847719
 Davidson iteration 16.4 (tol=0.0002616155258459103): -326.8313854877969 -326.7815582498002 -325.8769127380252 -325.0152503731554
 Davidson iteration 16.5 (tol=0.0002631618641042295): -326.8313855662656 -326.8005964133932 -326.2470836317568 -325.329066253238
 Davidson iteration 16.6 (tol=0.0002564815746118621): -326.831385641237 -326.8097665968214 -326.4605422777552 -325.5641239100666
 Davidson iteration 16.7 (tol=0.0002456770878358593): -326.8313857105018 -326.8148746241363 -326.584304179044 -325.8575468306166
 Davidson iteration 16.8 (tol=0.0002156821723359933): -326.8313857733629 -326.8180540294713 -326.6599917208368 -326.112441420808
 Davidson iteration 16.9 (tol=0.0003168997600048235): -326.8313858351075 -326.8203851976997 -326.7141469193739 -326.3377316652806
 Davidson iteration 17.0 (tol=0.000316899759995327): -326.8313858351068
 Davidson iteration 17.1 (tol=0.0001766346083586538): -326.8313858853476 -325.365918479793
 Davidson iteration 17.2 (tol=0.0001752425647726973): -326.8313859223417 -326.4258600650961 -324.5599691797308
 Davidson iteration 17.3 (tol=0.0001596984403773417): -326.8313859528604 -326.6656367994094 -325.003859562725 -324.4718397709711
 Davidson iteration 17.4 (tol=0.0001430312829883759): -326.8313859780181 -326.7399918281283 -325.6478211385117 -324.7105647310245
 Davidson iteration 17.5 (tol=0.000125397624722812): -326.8313859983456 -326.7711667396073 -326.1182569799687 -324.9683623897071
 Davidson iteration 17.6 (tol=0.0001090926090112448): -326.8313860136678 -326.7863736378836 -326.3815341269741 -325.3524271388201
 Davidson iteration 17.7 (tol=9.334613181603171e-05): -326.8313860250656 -326.7951582232355 -326.5272892029295 -325.7397053768698
 Elapsed time for davidson 4568.992171 (sec) 
 Elapsed time for diagonalization 4568.992232 (sec) 
 Elapsed time for mult 24.89221 (sec) 
 Energy = -326.8313860250677
 Elapsed time for measurement 0.010948 (sec) 
 truncated weight in carry-over = 2.523962537659408e-05
 Sample-based diagonalization: Energy = -326.8313860250677
 Sample-based diagonalization: density = [1.99994557515366,1.999976437882881,1.999990179777543,1.999833318028693,1.999799068354025,1.999624730125956,1.99984353237809,1.999916269073727,1.999870549344354,1.999586663765683,1.999685909931049,1.999867528194441,1.999694867903628,1.999804773770552,1.999996474226284,1.999862935967595,1.999935073761801,1.99995150919235,1.002571714147928,1.806122678869705,1.877173218635167,1.112865993593113,0.9999274578692081,1.97062762076211,1.940795239163196,1.849107279550616,1.979213045888638,0.9694472809407687,0.1543435904147825,0.003622496358971292,0.07412635043790478,0.9574938495630168,0.9795141622393361,0.1995446644331741,0.1255109264443794,0.0008070338561891874
 Sample-based diagonalization: carryover bitstrings =  000100000111110111111111111111111111 000010001111101110111111111111111111 000110000111110110111111111111111111 000000001111101111111111111111111111 000100000111101111111111111111111111 000010001111110110111111111111111111 ... 001101001101100110111111111111111111
```

here is an output with `SBD_TRADMODE`
```
 Elapsed time for helper construction 0.265809 (sec) 
 Elapsed time for init 0.001531 (sec) 
 Davidson iteration 0.0 (tol=0.5584049335932896): -326.2408941280003
 Davidson iteration 0.1 (tol=0.5273981009133736): -326.2727415524768 -326.2091473255696
 Davidson iteration 0.2 (tol=0.388204457239284): -326.4178430437822 -326.2181988615177 -326.0172122206592
 Davidson iteration 0.3 (tol=0.3497729713517192): -326.4713519238916 -326.3557367486418 -326.2181319105965 -326.0098087398594
 Davidson iteration 0.4 (tol=0.2808618773151046): -326.5832547466651 -326.3728581604678 -326.2488834558872 -326.2180261078601
 Davidson iteration 0.5 (tol=0.198733256359907): -326.6733247990203 -326.3827196556921 -326.254214642412 -326.2181261048361
 Davidson iteration 0.6 (tol=0.1424807026850615): -326.717065103882 -326.4076843308267 -326.2744276287287 -326.218708469311
 Davidson iteration 0.7 (tol=0.09859219297260297): -326.7372370812941 -326.4505003459511 -326.3130582992134 -326.2191884205282
 Davidson iteration 0.8 (tol=0.07931361945032463): -326.7479252638605 -326.5287199295561 -326.3509495832776 -326.2195689809966
 Davidson iteration 0.9 (tol=0.07419503855500596): -326.75528163037 -326.6045188660332 -326.3840757736334 -326.2207429243194
 Davidson iteration 1.0 (tol=0.07419503855500859): -326.7552816303472
 Davidson iteration 1.1 (tol=0.05140028195415903): -326.7588550928672 -325.6476554008519
 Davidson iteration 1.2 (tol=0.0670041635116501): -326.7633688330964 -326.5263805941568 -324.6727448376425
 Davidson iteration 1.3 (tol=0.08017598157450892): -326.7699572582404 -326.6839862147834 -325.3343254154428 -324.5798322956593
 Davidson iteration 1.4 (tol=0.07117575513543181): -326.7778571648172 -326.7243090870052 -325.8671417119532 -324.8959453820922
 Davidson iteration 1.5 (tol=0.05337071676127358): -326.782742615209 -326.7355446874851 -326.1993208165532 -325.3611160235101
 Davidson iteration 1.6 (tol=0.04595740421107942): -326.7858027869378 -326.7410707518957 -326.4419082138867 -325.7038866073128
 Davidson iteration 1.7 (tol=0.05131953226046268): -326.7888658822574 -326.746403505193 -326.6066017205891 -326.0200075655171
 Davidson iteration 1.8 (tol=0.06757810357598019): -326.7938282009554 -326.7555173758349 -326.6943962429306 -326.2344352680985
 Davidson iteration 1.9 (tol=0.05841650860721943): -326.7983220873277 -326.7631261326055 -326.7163566232279 -326.3337459669934
 Davidson iteration 2.0 (tol=0.05841650860727956): -326.7983220873355
 Davidson iteration 2.1 (tol=0.04343399920585904): -326.8005272628686 -325.6989202154597
 Davidson iteration 2.2 (tol=0.04548367796850045): -326.8032473056741 -326.5305444922893 -324.640427093645
 Davidson iteration 2.3 (tol=0.05122468478780325): -326.8061931393706 -326.6930326491574 -325.6960109727837 -324.3988481026421
 Davidson iteration 2.4 (tol=0.05545919784909142): -326.8097182236088 -326.7492576997347 -326.0262730170614 -325.2668972080129
 Davidson iteration 2.5 (tol=0.05462369672609727): -326.8134696315619 -326.7716327931915 -326.2773427681497 -325.6737432736398
 Davidson iteration 2.6 (tol=0.04528226600676757): -326.8164141825677 -326.7807957212207 -326.4447332692444 -325.8162148582715
 Davidson iteration 2.7 (tol=0.04121761600391612): -326.8186074125443 -326.7856988480205 -326.5643711588443 -326.0240825954571
 Davidson iteration 2.8 (tol=0.03267148913820877): -326.8202079888685 -326.7887992206051 -326.6406668136262 -326.1975608276073
 Davidson iteration 2.9 (tol=0.03376459665189684): -326.8212990277959 -326.7909271694315 -326.68825721939 -326.334715525066
 Davidson iteration 3.0 (tol=0.03376459665191736): -326.8212990277918
 Davidson iteration 3.1 (tol=0.02138891533081055): -326.8219515957762 -325.5264081466293
 Davidson iteration 3.2 (tol=0.02443380290512191): -326.8226057277935 -326.522138514918 -324.5506516989606
 Davidson iteration 3.3 (tol=0.02438562750284775): -326.8232937352757 -326.7061578789307 -325.1908004234983 -324.4378067125765
 Davidson iteration 3.4 (tol=0.02438881517101366): -326.8240048975371 -326.7628725044524 -325.8261803656382 -324.7892325274544
 Davidson iteration 3.5 (tol=0.02119646535646047): -326.8246376890716 -326.7841438355145 -326.2030074578869 -325.1827134367816
 Davidson iteration 3.6 (tol=0.01728134503985551): -326.8250842727468 -326.793191805236 -326.4088940520842 -325.5626201196674
 Davidson iteration 3.7 (tol=0.01410112168643186): -326.8253991665229 -326.7979418416892 -326.5426613379253 -325.9693329233559
 Davidson iteration 3.8 (tol=0.01108794359857416): -326.8256092585518 -326.8007463846032 -326.62805570437 -326.2069175940953
 Davidson iteration 3.9 (tol=0.01045088626713294): -326.8257148729838 -326.8021284371927 -326.6713446622979 -326.3308399553111
 Davidson iteration 4.0 (tol=0.01045088626715063): -326.825714872988
 Davidson iteration 4.1 (tol=0.007894392447088807): -326.8257857500774 -325.6993830915507
 Davidson iteration 4.2 (tol=0.007818293586323672): -326.8258689303879 -326.5388388881354 -324.5587491350623
 Davidson iteration 4.3 (tol=0.007252443000638754): -326.8259387136601 -326.693009463162 -325.6039145888167 -324.2970495020753
 Davidson iteration 4.4 (tol=0.006529475887386034): -326.8259954031722 -326.746820502964 -325.9677562844869 -325.1756535163057
 Davidson iteration 4.5 (tol=0.005658501932827411): -326.826039616292 -326.7708892074642 -326.2458927831289 -325.5695016728428
 Davidson iteration 4.6 (tol=0.004576097799135509): -326.8260699350485 -326.7823625383714 -326.4263298408509 -325.769089783662
 Davidson iteration 4.7 (tol=0.004040609801457541): -326.8260911278986 -326.7892660264429 -326.5524871809623 -326.0083001634078
 Davidson iteration 4.8 (tol=0.003266057081586567): -326.8261067159673 -326.7939472959982 -326.6332196762434 -326.1844414150623
 Davidson iteration 4.9 (tol=0.004096062329645446): -326.8261191644589 -326.7978139424191 -326.6894862032323 -326.3384760209339
 Davidson iteration 5.0 (tol=0.004096062329651331): -326.8261191644582
 Davidson iteration 5.1 (tol=0.002407013502181484): -326.8261280282292 -325.4083613867979
 Davidson iteration 5.2 (tol=0.00254498387856309): -326.8261355089276 -326.4646348897735 -324.5427765127401
 Davidson iteration 5.3 (tol=0.002432631222165158): -326.8261425001571 -326.6819027319738 -325.1180804801385 -324.440729064627
 Davidson iteration 5.4 (tol=0.002425569683854807): -326.8261493415111 -326.7521648868918 -325.7791675412254 -324.7482045415022
 Davidson iteration 5.5 (tol=0.002281262878099328): -326.8261559046801 -326.7806388777461 -326.189693122031 -325.1002011312816
 Davidson iteration 5.6 (tol=0.002115527784771208): -326.826161616537 -326.7940841935528 -326.4157419756718 -325.5105172929906
 Davidson iteration 5.7 (tol=0.001991317073709009): -326.826166781498 -326.8017661301923 -326.553412997124 -325.9238367165777
 Davidson iteration 5.8 (tol=0.00187238536940742): -326.8261715823226 -326.8068376940398 -326.6409563857352 -326.1880544643249
 Davidson iteration 5.9 (tol=0.002289208022783787): -326.8261753111692 -326.8099177170562 -326.6889412571651 -326.3351250781727
 Davidson iteration 6.0 (tol=0.002289208022725868): -326.8261753111664
 Davidson iteration 6.1 (tol=0.001584230506308963): -326.8261783814856 -325.6040818201074
 Davidson iteration 6.2 (tol=0.001603571841705376): -326.8261816924779 -326.5202459484846 -324.5221457788356
 Davidson iteration 6.3 (tol=0.001691713113326594): -326.8261849582121 -326.7037647836821 -325.5616376453633 -324.2945705674285
 Davidson iteration 6.4 (tol=0.001896336158437876): -326.8261887366872 -326.7671944057417 -325.9621126658236 -325.1424579719683
 Davidson iteration 6.5 (tol=0.002266190682748636): -326.8261938370149 -326.7954637469631 -326.2754050514819 -325.5159917695902
 Davidson iteration 6.6 (tol=0.002800275281502448): -326.8262011720957 -326.8092384972177 -326.4683797216223 -325.7361143456834
 Davidson iteration 6.7 (tol=0.00399027365347491): -326.8262139200324 -326.8171499830616 -326.5893889603415 -325.9924703508534
 Davidson iteration 6.8 (tol=0.006333293070394772): -326.8262437739975 -326.8220344894751 -326.6617266558742 -326.1867744561258
 Davidson iteration 6.9 (tol=0.02651030646236624): -326.8264079636891 -326.8251772284255 -326.7095414498882 -326.3549001889718
 Davidson iteration 7.0 (tol=0.02651030646236862): -326.8264079636918
 Davidson iteration 7.1 (tol=0.01446067324456974): -326.8267563160144 -325.327429534191
 Davidson iteration 7.2 (tol=0.01377810885837969): -326.8269973817596 -326.40160083392 -324.5465643244038
 Davidson iteration 7.3 (tol=0.01251143881894542): -326.8271925062136 -326.6543032720368 -325.0829271570734 -324.4551361789343
 Davidson iteration 7.4 (tol=0.01187193743617107): -326.8273632629166 -326.7371997116032 -325.7320576599345 -324.737637624852
 Davidson iteration 7.5 (tol=0.01119734970011098): -326.8275206286926 -326.772812088118 -326.1711402694953 -325.0693865290228
 Davidson iteration 7.6 (tol=0.01073603046148396): -326.8276620344819 -326.7904846464231 -326.413301298108 -325.4683435290239
 Davidson iteration 7.7 (tol=0.01059707262251217): -326.8277986944636 -326.8010617876563 -326.554468138736 -325.8759405182385
 Davidson iteration 7.8 (tol=0.01035118653364241): -326.8279408521935 -326.8082345886612 -326.6411521484723 -326.1557526962051
 Davidson iteration 7.9 (tol=0.01424410315562168): -326.8280674994808 -326.8126108252224 -326.6906090767121 -326.3258560144757
 Davidson iteration 8.0 (tol=0.01424410315562347): -326.8280674994814
 Davidson iteration 8.1 (tol=0.00991276002939729): -326.8281867370978 -325.6278787495811
 Davidson iteration 8.2 (tol=0.01062081243075205): -326.8283251541746 -326.5562421878427 -324.4955503643819
 Davidson iteration 8.3 (tol=0.01120940127905132): -326.8284669966865 -326.7209937024257 -325.5208979039675 -324.2735384259726
 Davidson iteration 8.4 (tol=0.01205763137443869): -326.8286254844582 -326.7750708143419 -325.9437434499411 -325.1078141340993
 Davidson iteration 8.5 (tol=0.013095891576845): -326.8288121739156 -326.7983003042265 -326.272516957372 -325.4624100529032
 Davidson iteration 8.6 (tol=0.01386737126028106): -326.829021287653 -326.8094243507331 -326.468707642951 -325.6854410451898
 Davidson iteration 8.7 (tol=0.01491542446988013): -326.8292562956503 -326.8155879618495 -326.5881834937337 -325.9536593745557
 Davidson iteration 8.8 (tol=0.01456056715222466): -326.8295167703953 -326.8193076116092 -326.6611381001463 -326.1670045090902
 Davidson iteration 8.9 (tol=0.02218950942314267): -326.8298158897269 -326.8218087213784 -326.7117040126858 -326.354032620387
 Davidson iteration 9.0 (tol=0.02218950942314603): -326.8298158897228
 Davidson iteration 9.1 (tol=0.01233312163245378): -326.8300642113787 -325.3679976886928
 Davidson iteration 9.2 (tol=0.01204411935990142): -326.8302446214724 -326.426908290613 -324.5477069264207
 Davidson iteration 9.3 (tol=0.01086098868439595): -326.8303912579059 -326.6645387139398 -325.061683621425 -324.4570171439842
 Davidson iteration 9.4 (tol=0.009827007168369954): -326.8305125052748 -326.7400407480273 -325.7046356399994 -324.728892225862
 Davidson iteration 9.5 (tol=0.008632993269792583): -326.8306117850295 -326.7718919269948 -326.1503738758767 -325.0419015920268
 Davidson iteration 9.6 (tol=0.007557333110888706): -326.8306874950127 -326.7874505844544 -326.3971143828325 -325.4298055131591
 Davidson iteration 9.7 (tol=0.006638170463940518): -326.8307464028443 -326.7966335455417 -326.5389830334167 -325.8278275491376
 Davidson iteration 9.8 (tol=0.005460990566546598): -326.8307935109619 -326.8026925900743 -326.6254330778983 -326.1159100676932
 Davidson iteration 9.9 (tol=0.006862474709272067): -326.8308251630896 -326.8064128970836 -326.6783209945635 -326.3079693992182
 Davidson iteration 10.0 (tol=0.006862474709224072): -326.8308251630857
 Davidson iteration 10.1 (tol=0.004968448980168143): -326.8308536169053 -325.6525053178546
 Davidson iteration 10.2 (tol=0.005727890727516652): -326.8308910842559 -326.584135235384 -324.4729683203631
 Davidson iteration 10.3 (tol=0.006071252112425417): -326.8309317328756 -326.7330617251665 -325.4849944160765 -324.2543184592753
 Davidson iteration 10.4 (tol=0.006322401678805121): -326.8309761699928 -326.780257187207 -325.9245482292429 -325.0777007393423
 Davidson iteration 10.5 (tol=0.006418348849862786): -326.8310235254464 -326.8000627295974 -326.2656162151779 -325.4178644119741
 Davidson iteration 10.6 (tol=0.006293143118906995): -326.8310694260107 -326.8096062650064 -326.4667385313275 -325.6460382050923
 Davidson iteration 10.7 (tol=0.006159568751851669): -326.8311129245533 -326.8149591462747 -326.5870803160669 -325.9224194216236
 Davidson iteration 10.8 (tol=0.005499205957664407): -326.8311533792615 -326.818290282307 -326.6606646941643 -326.1486102936726
 Davidson iteration 10.9 (tol=0.007938205266401193): -326.8311933093007 -326.8206789313792 -326.7125000308379 -326.3488142195183
 Davidson iteration 11.0 (tol=0.007938205266410137): -326.831193309312
 Davidson iteration 11.1 (tol=0.004426110267180962): -326.8312250629221 -325.3686793781839
 Davidson iteration 11.2 (tol=0.004348867633352149): -326.8312483298745 -326.4272380714915 -324.5496608215708
 Davidson iteration 11.3 (tol=0.003926281232965237): -326.8312672790134 -326.6647017210259 -325.0424774809977 -324.4601071643591
 Davidson iteration 11.4 (tol=0.003523630201997803): -326.8312828090483 -326.7394714134215 -325.6846671709913 -324.7209732397241
 Davidson iteration 11.5 (tol=0.003080535679690207): -326.8312953586895 -326.7710417251116 -326.1390245477113 -325.0175229126284
 Davidson iteration 11.6 (tol=0.002673877617127081): -326.8313047971409 -326.7864431795721 -326.3907108720674 -325.4030924899547
 Davidson iteration 11.7 (tol=0.002308892714029646): -326.8313119199198 -326.7954434339139 -326.5332822402879 -325.7966888040519
 Davidson iteration 11.8 (tol=0.001839800825237995): -326.8313174006445 -326.8012385986557 -326.6195179037617 -326.0919836630983
 Davidson iteration 11.9 (tol=0.002314769080805312): -326.8313209505 -326.8047513805308 -326.6739567124046 -326.2982286020277
 Davidson iteration 12.0 (tol=0.002314769080831606): -326.831320950488
 Davidson iteration 12.1 (tol=0.001683968047855768): -326.83132416869 -325.6389457598092
 Davidson iteration 12.2 (tol=0.001986496381937429): -326.8313285362947 -326.5860228487903 -324.459489124442
 Davidson iteration 12.3 (tol=0.002118479461276679): -326.831333385249 -326.7344141836778 -325.4552227489585 -324.2472727351758
 Davidson iteration 12.4 (tol=0.002201943442659391): -326.8313387284319 -326.7809330167801 -325.9062799420901 -325.0539462111964
 Davidson iteration 12.5 (tol=0.002219280092175037): -326.8313443734103 -326.800281709183 -326.2586043027408 -325.3828272059568
 Davidson iteration 12.6 (tol=0.002163385154256533): -326.8313497761449 -326.8096146805414 -326.4645413793997 -325.613596234041
 Davidson iteration 12.7 (tol=0.002090606651635489): -326.8313548073706 -326.8148395468145 -326.5861486065476 -325.8966400823351
 Davidson iteration 12.8 (tol=0.001848672529147346): -326.8313594099801 -326.8180993826717 -326.6605556623723 -326.1341041233269
 Davidson iteration 12.9 (tol=0.002684683405775689): -326.8313639243252 -326.8204676565496 -326.7133840213648 -326.3446125578338
 Davidson iteration 13.0 (tol=0.002684683405760038): -326.8313639243281
 Davidson iteration 13.1 (tol=0.00149652111389192): -326.8313675449895 -325.3666060339796
 Davidson iteration 13.2 (tol=0.00147627200492255): -326.8313702029766 -326.4263391345108 -324.5529931296331
 Davidson iteration 13.3 (tol=0.001338450178668994): -326.8313723799755 -326.6649078285193 -325.0272334934452 -324.4641288224078
 Davidson iteration 13.4 (tol=0.001200356305964475): -326.8313741688818 -326.7395396880834 -325.6698474935736 -324.7161256931164
 Davidson iteration 13.5 (tol=0.001050972067100083): -326.8313756152929 -326.7709716865609 -326.1308201346263 -324.9980402821917
 Davidson iteration 13.6 (tol=0.0009127081146999037): -326.8313767043014 -326.7862970377458 -326.3869689816172 -325.382819372718
 Davidson iteration 13.7 (tol=0.0007843678621576052): -326.8313775198911 -326.7951966497911 -326.5305506911856 -325.7735210583299
 Davidson iteration 13.8 (tol=0.0006177990240089032): -326.831378141369 -326.8008671637189 -326.6170121927141 -326.0748615620674
 Davidson iteration 13.9 (tol=0.0007837555069094398): -326.8313785416305 -326.804287979215 -326.6725537527144 -326.2916892672791
 Davidson iteration 14.0 (tol=0.0007837555068564781): -326.8313785416344
 Davidson iteration 14.1 (tol=0.0005697460966853757): -326.8313789077527 -325.629105094266
 Davidson iteration 14.2 (tol=0.0006793767266081498): -326.8313794095401 -326.5860681771376 -324.4491605559806
 Davidson iteration 14.3 (tol=0.0007275586755016729): -326.8313799731214 -326.7349697399018 -325.4300655723303 -324.2434812163768
 Davidson iteration 14.4 (tol=0.0007563718814021683): -326.8313805984304 -326.7812973904416 -325.8905182543092 -325.0335527850962
 Davidson iteration 14.5 (tol=0.0007614449535919533): -326.8313812587947 -326.800458683365 -326.2524586968424 -325.3538422145022
 Davidson iteration 14.6 (tol=0.0007420411466831559): -326.8313818900363 -326.8096928796618 -326.462444290145 -325.5867480464556
 Davidson iteration 14.7 (tol=0.0007132871129740408): -326.8313824749818 -326.8148471189088 -326.5851959128096 -325.8753710773218
 Davidson iteration 14.8 (tol=0.0006281948604979734): -326.8313830075086 -326.8180593530279 -326.660310647401 -326.1223138546727
 Davidson iteration 14.9 (tol=0.0009185741620302342): -326.8313835303629 -326.8204069797044 -326.7139016960695 -326.3410012062939
 Davidson iteration 15.0 (tol=0.0009185741620342883): -326.831383530363
 Davidson iteration 15.1 (tol=0.0005120057687082147): -326.8313839532694 -325.3661281683563
 Davidson iteration 15.2 (tol=0.0005067445192115045): -326.8313842642899 -326.4261477388512 -324.5563973160699
 Davidson iteration 15.3 (tol=0.0004608898834595038): -326.8313845201702 -326.6653661066206 -325.014657572067 -324.467987808616
 Davidson iteration 15.4 (tol=0.0004131029334642481): -326.8313847308916 -326.7398177190576 -325.6579078440446 -324.7128394169526
 Davidson iteration 15.5 (tol=0.0003619878855852615): -326.8313849012587 -326.771091875787 -326.1240804915914 -324.9819833732147
 Davidson iteration 15.6 (tol=0.0003146304554459176): -326.8313850296145 -326.7863398807478 -326.3840560428997 -325.3663743335218
 Davidson iteration 15.7 (tol=0.0002696657148730309): -326.8313851253383 -326.795165051588 -326.5287369913747 -325.7550485539862
 Davidson iteration 15.8 (tol=0.0002108581368080578): -326.8313851979184 -326.8007606987852 -326.6155722422914 -326.0613114489375
 Davidson iteration 15.9 (tol=0.0002693031347559761): -326.8313852445756 -326.8041313904845 -326.6719406848798 -326.2864357238008
 Davidson iteration 16.0 (tol=0.0002693031347272566): -326.8313852445747
 Davidson iteration 16.1 (tol=0.0001955007504026613): -326.8313852875202 -325.6222797951286
 Davidson iteration 16.2 (tol=0.0002346116767919264): -326.8313853466397 -326.5860133414091 -324.44039014484
 Davidson iteration 16.3 (tol=0.000251800895887373): -326.8313854134344 -326.7353572988375 -325.4081998408335 -324.2406741847719
 Davidson iteration 16.4 (tol=0.0002616155258459103): -326.8313854877969 -326.7815582498002 -325.8769127380252 -325.0152503731554
 Davidson iteration 16.5 (tol=0.0002631618641042295): -326.8313855662656 -326.8005964133932 -326.2470836317568 -325.329066253238
 Davidson iteration 16.6 (tol=0.0002564815746118621): -326.831385641237 -326.8097665968214 -326.4605422777552 -325.5641239100666
 Davidson iteration 16.7 (tol=0.0002456770878358593): -326.8313857105018 -326.8148746241363 -326.584304179044 -325.8575468306166
 Davidson iteration 16.8 (tol=0.0002156821723359933): -326.8313857733629 -326.8180540294713 -326.6599917208368 -326.112441420808
 Davidson iteration 16.9 (tol=0.0003168997600048235): -326.8313858351075 -326.8203851976997 -326.7141469193739 -326.3377316652806
 Davidson iteration 17.0 (tol=0.000316899759995327): -326.8313858351068
 Davidson iteration 17.1 (tol=0.0001766346083586538): -326.8313858853476 -325.365918479793
 Davidson iteration 17.2 (tol=0.0001752425647726973): -326.8313859223417 -326.4258600650961 -324.5599691797308
 Davidson iteration 17.3 (tol=0.0001596984403773417): -326.8313859528604 -326.6656367994094 -325.003859562725 -324.4718397709711
 Davidson iteration 17.4 (tol=0.0001430312829883759): -326.8313859780181 -326.7399918281283 -325.6478211385117 -324.7105647310245
 Davidson iteration 17.5 (tol=0.000125397624722812): -326.8313859983456 -326.7711667396073 -326.1182569799687 -324.9683623897071
 Davidson iteration 17.6 (tol=0.0001090926090112448): -326.8313860136678 -326.7863736378836 -326.3815341269741 -325.3524271388201
 Davidson iteration 17.7 (tol=9.334613181603171e-05): -326.8313860250656 -326.7951582232355 -326.5272892029295 -325.7397053768698
 Elapsed time for davidson 2623.80041 (sec) 
 Elapsed time for diagonalization 2623.800487 (sec) 
 Elapsed time for mult 13.88541 (sec) 
 Energy = -326.8313860250677
 Elapsed time for measurement 0.01328 (sec) 
 truncated weight in carry-over = 2.523962537659408e-05
 Sample-based diagonalization: Energy = -326.8313860250677
 Sample-based diagonalization: density = [1.99994557515366,1.999976437882881,1.999990179777544,1.999833318028693,1.999799068354025,1.999624730125956,1.99984353237809,1.999916269073726,1.999870549344354,1.999586663765683,1.99968590993105,1.999867528194441,1.999694867903629,1.999804773770552,1.999996474226284,1.999862935967595,1.999935073761801,1.99995150919235,1.002571714147928,1.806122678869705,1.877173218635167,1.112865993593113,0.9999274578692081,1.97062762076211,1.940795239163196,1.849107279550616,1.979213045888638,0.9694472809407686,0.1543435904147825,0.003622496358971292,0.07412635043790478,0.9574938495630168,0.9795141622393361,0.1995446644331741,0.1255109264443795,0.0008070338561891874
 Sample-based diagonalization: carryover bitstrings =  000100000111110111111111111111111111 000010001111101110111111111111111111 000110000111110110111111111111111111 000000001111101111111111111111111111 000100000111101111111111111111111111 000010001111110110111111111111111111 ... 001101001101100110111111111111111111
```

I think it runs correctly and get almost x2 speedup with the optimization in `SBD_TRADMODE`

walkup committed:
added a set of bitstring files to provide a range of inputs described in the README

walkup committed:
added inputs for H2O with the cc-pvdz basis set, with up to 1.52e9 determinants

walkup commented:
Thanks for the testing on Fugaku.

There is one additional set of runtime options that would be very good to test on Fugaku, using the same fcidump and AlphaDets inputs and the same number of nodes as the previous jobs.

Please try : one MPI rank per node with 48 OpenMP threads ... so 96 MPI ranks for a 96-node job.

The options below are the ones that differ from the previous jobs, all other options are not changed.

       --block 30 --iteration 10   \
       --adet_comm_size 2 --bdet_comm_size 8 --task_comm_size 1 \
       --shuffle 1 --rdm 1 

When using one MPI rank per node with 48 threads, one can increase the --block parameter, which will also increase the memory requirement per MPI rank.  Our experience has been that for some  cases, the Davidson algorithm converges in fewer iterations when the size of the Davidson subspace is allowed to expand beyond 10.  When using a lot of threads, it might be better to use more ranks to partition the "b" index (--bdet_comm_size) than the "a" index (--adet_comm_size), because OpenMP is used exclusively over the "a" index.  Turning on --shuffle 1 can improve load balance, particularly when not using the task-based parallelization scheme.  There were code changes to improve threading in the part of the code that computes reduced density matrix values, and to see the impact we need to have --rdm 1.

walkup committed:
updated reference outputs for H2O

walkup committed:
minor fix to error message from davidson.h

t-sirakawa commented:
Thank you very much for the great work.
Although the dataset is not exactly the same, I ran a 2Fe–2S case with the subspace dimension matched to your suggested setting.
I also followed your recommended configuration of 48 threads per node × 96 nodes, and I would like to share the results below.

- Original code:
```
 Elapsed time for diagonalization 4819.953217 (sec) 
 Elapsed time for mult 74.35011999999999 (sec) 
```
- Tuned code (clang mode)
```
 Elapsed time for diagonalization 2021.658666 (sec) 
 Elapsed time for mult 23.389497 (sec) 
```
- Tuned code (trad mode with `-DSBD_TRADMODE` option)
```
 Elapsed time for diagonalization 1602.330467 (sec) 
 Elapsed time for mult 18.568589 (sec) 
```
As you can see, using the configuration you proposed results in almost a **3× speed-up**, which is fantastic!
I also confirmed that everything runs without issues, so I will go ahead and merge this pull request.

Thank you again -- really appreciate all the work from the IBM Research Collaborators!
